java/vector/src/main/java/org/apache/arrow/vector/BitVectorHelper.java

```diff
@@ -319,4 +321,73 @@ static void setBitMaskedByte(ArrowBuf data, int byteIndex, byte bitMask) {
     currentByte |= bitMask;
     data.setByte(byteIndex, currentByte);
   }
+
+  /**
+   * Concat two validity buffers.
+   * @param input1 the first validity buffer.
+   * @param numBits1 the number of bits in the first validity buffer.
+   * @param input2 the second validity buffer.
+   * @param numBits2 the number of bits in the second validity buffer.
+   * @param output the ouput validity buffer. It can be the same one as the first input.
+   */
+  public static void concatBits(ArrowBuf input1, int numBits1, ArrowBuf input2, int numBits2, ArrowBuf output) {
+    int numBytes1 = DataSizeRoundingUtil.divideBy8Ceil(numBits1);
+    int numBytes2 = DataSizeRoundingUtil.divideBy8Ceil(numBits2);
+    int numBytesOut = DataSizeRoundingUtil.divideBy8Ceil(numBits1 + numBits2);
+
+    if (BoundsChecking.BOUNDS_CHECKING_ENABLED) {
+      output.checkBytes(0, numBytesOut);
+    }
+
+    // copy the first bit set
+    if (input1 != output) {
+      PlatformDependent.copyMemory(input1.memoryAddress(), output.memoryAddress(), numBytes1);
+    }
+    output.setZero(numBytes1, numBytesOut);
+
+    if (numBits1 % 8 == 0) {
+      PlatformDependent.copyMemory(input2.memoryAddress(), output.memoryAddress() + numBytes1, numBytes2);
+      return;
+    }
+
+    // the number of bits to fill a full byte after the first input is processed
+    int numBitsToFill = 8 - (numBits1 % 8);
```

Comment:
mask instead of mod, don't we have a utility method for this?

Reply:
Revised. Thanks.